`ref` fields and `scoped`. Modify pages on struct creation, variable declaration, and ref struct. (#30912)
dotnet/docs: docs/csharp/language-reference/statements/declarations.md

---
title: "Declaration statements - C# reference"
description: "Declaration statements, including `var`, `ref` locals, and `ref` fields - C# reference"
ms.date: 09/15/2022
f1_keywords: 
  - "var"
  - "var_CSharpKeyword"
helpviewer_keywords: 
  - "var keyword [C#]"
---
# Declaration statements

A *declaration statement* declares a new variable, and optionally, initializes it. All variables have declared type. You can learn more about types in the article on the [.NET type system](../../../standard/base-types/common-type-system.md).

## Implicitly typed local variables

> Beginning with C# 3, variables that are declared at method scope can have an implicit "type" `var`. An implicitly typed local variable is strongly typed as if you had declared the type yourself, but the compiler determines the type. The following two declarations of `i` are functionally equivalent:

Reviewer: ```suggestion
Beginning with C# 3, variables that are declared at method scope can have an implicit "type" `var`. An implicitly typed local variable is strongly typed as if you had declared the type yourself, but the compiler determines the type. The following two declarations of `a` and `b` are functionally equivalent:
```

```csharp

> var i = 10; // Implicitly typed.
> int i = 10; // Explicitly typed.

Reviewer: ```suggestion
var a = 10; // Implicitly typed.
int b = 10; // Explicitly typed.
```

```

The preceding assignment declares `p` as a local variable. Its initial value is copied from reading the value returned by `GetContactInformation`. Any future assignments to `p` won't change the value of the variable returned by `GetContactInformation`. The variable `p` is no longer an alias to the variable returned.

You declare a *ref* variable to copy the alias to the original value. In the following assignment, `p` is an alias to the variable returned from `GetContactInformation`.

```csharp
ref Person p = ref contacts.GetContactInformation("Brandie", "Best");
```

Subsequent usage of `p` is the same as using the variable returned by `GetContactInformation` because `p` is an alias for that variable. Changes to `p` also change the variable returned from `GetContactInformation`.

You can access a value by reference in the same way. In some cases, accessing a value by reference increases performance by avoiding a potentially expensive copy operation. For example, the following statement shows how one can define a ref local value that is used to reference a value.

```csharp
ref VeryLargeStruct reflocal = ref veryLargeStruct;
```

> The `ref` keyword is used both before the local variable declaration *and* before the value in the second example. Failure to include both `ref` keywords in the variable declaration and assignment in both examples results in compiler error CS8172, "Can't initialize a by-reference variable with a value."

Reviewer: Could we link to compiler error CS8172 here?

Author: not yet. We don't have help on that one yet.

```csharp
ref VeryLargeStruct reflocal = ref veryLargeStruct; // initialization
refLocal = ref anotherVeryLargeStruct; // reassigned, refLocal refers to different storage.
```

>  Ref local variables must still be initialized when they're declared.

Reviewer: ```suggestion
Ref local variables must still be initialized when they're declared.
```
